core/cache/charmconfigwatcher.go

```diff
@@ -0,0 +1,273 @@
+	initComplete chan struct{}
+
+	unitName   string
+	appName    string
+	branchName string
+
+	masterSettings map[string]interface{}
+	branchDeltas   settings.ItemChanges
+	configHash     string
+}
+
+// newUnitConfigWatcher returns a new watcher for the input config keys
+// with a baseline hash of their config values from the input hash cache.
+// As per the cache requirements, hashes are only generated from sorted keys.
+func newCharmConfigWatcher(cfg charmConfigWatcherConfig) (*CharmConfigWatcher, error) {
+	w := &CharmConfigWatcher{
+		notifyWatcherBase: newNotifyWatcherBase(),
+		initComplete:      make(chan struct{}),
+		unitName:          cfg.unitName,
+		appName:           cfg.appName,
+	}
+
+	deregister := cfg.res.registerWorker(w)
+
+	multi := cfg.hub.NewMultiplexer()
+	multi.Add(cfg.appConfigChangeTopic, w.appConfigChanged)
+	multi.Add(cfg.branchChangeTopic, w.branchChanged)
+	multi.Add(cfg.branchRemoveTopic, w.branchRemoved)
+
+	w.tomb.Go(func() error {
+		<-w.tomb.Dying()
+		multi.Unsubscribe()
+		deregister()
+		return nil
+	})
+
+	if err := w.init(cfg.model); err != nil {
+		_ = w.Stop()
+		return nil, errors.Trace(err)
+	}
+	return w, nil
+}
+
+// init determines baseline master configuration, branch tracking settings
+// and the configuration hash for the watcher's unit.
+// It then closes the init channel to indicate the watcher is operational.
+func (w *CharmConfigWatcher) init(model charmConfigModel) error {
+	app, err := model.Application(w.appName)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	w.masterSettings = app.Config()
+
+	for _, b := range model.Branches() {
+		if w.isTracking(b) {
+			w.branchName = b.Name()
+			w.branchDeltas = b.AppConfig(w.appName)
+			break
+		}
+	}
+
+	if _, err := w.setConfigHash(); err != nil {
+		return errors.Trace(err)
+	}
+
+	close(w.initComplete)
+	return nil
+}
+
+// appConfigChanged is called when a message is received indicating changed
+// application master charm configuration.
+func (w *CharmConfigWatcher) appConfigChanged(_ string, msg interface{}) {
+	if !w.waitInitOrDying() {
+		return
+	}
+
+	hashCache, ok := msg.(*hashCache)
+	if !ok {
+		logger.Errorf("programming error; application config message was not of expected type, *hashCache")
+		return
+	}
+
+	w.masterSettings = hashCache.config
+	w.checkConfig()
+}
+
+// branchChanged is called when we receive a message to say that a branch has
+// been updated in the cache.
+// If it is the branch that this watcher's unit is tracking,
+// check if the latest config delta warrants a notification.
+func (w *CharmConfigWatcher) branchChanged(_ string, msg interface{}) {
+	if !w.waitInitOrDying() {
+		return
+	}
+
+	b, okUnit := msg.(Branch)
+	if !okUnit {
+		logger.Errorf("programming error; branch change message was not of expected type, Branch")
+		return
+	}
+
+	// If we do not know whether we are tracking this branch, find out.
+	if w.branchName == "" {
+		if w.isTracking(b) {
+			w.branchName = b.Name()
+		}
```

Comment:
don't we want an else return in here too?

Reply:
Big time. Stand by.